Issue policy and template tweaks

dbczumar opened the pull request:
## What changes are proposed in this pull request?

* Revamp "feature request" template to prompt filers for specific motivation questions and high-level implementation details (if possible)
* Revamp Issue Policy to address the lifecycle for each issue type
* Add "Willingness to contribute" sections to the feature request, bug report, and documentation fix templates

## How is this patch tested?

Manually - I recommend viewing the associated files in a markdown renderer (e.g., open the file in the GitHub file browser).

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Model Registry
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

dbczumar committed:
Initial

dbczumar committed:
Tweaks

dbczumar committed:
Spacing

dbczumar committed:
Bug fix

dbczumar committed:
Lang

dbczumar committed:
Unbold

dbczumar committed:
Doc issues

dbczumar committed:
Tweak

dbczumar committed:
Installation

dbczumar committed:
Issue policy

dbczumar committed:
Ordering

dbczumar committed:
Update guidelines

dbczumar commented on .github/ISSUE_TEMPLATE/feature_request_template.md:
```diff
@@ -4,29 +4,81 @@ about: Use this template for feature and enhancement proposals.
+- [ ] Artifacts
+- [ ] Models
+- [ ] Model Registry
+- [ ] Scoring
+- [ ] Serving
```
@sueann Can we dedupe `Scoring` and `Serving` here and in the pull request template?

aarondav replied:
Maybe also API and REST-API? Dunno how a user would choose between those.

sueann replied:
sounds good - i think we want to just keep "Scoring"

dbczumar replied:
Thanks for the input! I've consolidated `Serving` and `Scoring` to just `Scoring`. I've consolidated `API` and `REST-API` as just `API`.

aarondav commented on .github/ISSUE_TEMPLATE/bug_report_template.md:
Do I want a _thorough_ response? Quick? Some other words?

dbczumar replied:
I've changed this to `timely and thorough`. More than happy to iterate here though!

dbczumar replied:
^ I applied these changes to the PR template as well

aarondav commented on .github/ISSUE_TEMPLATE/bug_report_template.md:
Probably we can leave out the Explanation bit. Just having "No" here is already 🤨

dbczumar replied:
Done!

aarondav commented on .github/ISSUE_TEMPLATE/feature_request_template.md:
```diff
@@ -4,29 +4,81 @@ about: Use this template for feature and enhancement proposals.
+
+- [ ] Yes. I think this feature can be introduced as an MLflow Plugin.
+- [ ] No. It does not make sense to structure this feature as an MLflow Plugin. **Explanation:**
+- [ ] No. I don’t think MLflow Plugins currently supports this type of feature. **Explanation:**
```
s/supports/support/?

dbczumar replied:
Done! Thanks!

dbczumar committed:
Address feedback

dbczumar committed:
Address

dbczumar committed:
Spacing

dbczumar committed:
Progress

dbczumar committed:
Format

dbczumar committed:
template update

dbczumar commented:
@aarondav @smurching @mateiz In response to feedback that the feature request template is too lengthy and technical, I've removed the sections related to plugins and `implementation impact` from the template. Instead, I've updated the `CONTRIBUTING.rst` document to provide guidelines about feature / patch design and information about structuring features as plugins, as well as an overview of the contribution process. The feature request template explicitly encourages users to consult the contributing guide. Another round of reviews would be greatly appreciated!

dbczumar commented on CONTRIBUTING.rst:
```diff
@@ -1,13 +1,105 @@
+MLflow committers actively triage and respond to GitHub issues. In general, we recommend waiting
+for feebdack from an MLflow committer or community member before proceeding to implement a feature
+or patch. This is particularly important for
+`significant changes <https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#write-designs-for-significant-changes>`_.
```
I tried using RST references (`:ref:`) to link to sections within this document. Unfortunately, GitHub doesn't seem to render RST references properly (see https://github.com/isaacs/github/issues/892), so I gave this a shot instead.

mateiz commented on .github/ISSUE_TEMPLATE/feature_request_template.md:
```diff
@@ -4,29 +4,48 @@ about: Use this template for feature and enhancement proposals.
+## Which MLflow component(s) does this feature affect?
+
+- [ ] UI
+- [ ] CLI
```
Some people might not know what CLI stands for so maybe spell it out

dbczumar replied:
Changed this to read `Command Line Interface` and also made the corresponding change to the pull request template. Thanks for pointing this out!

dbczumar committed:
Address comments

dbczumar committed:
Whitespace fixes for contributing guide

smurching commented on .github/ISSUE_TEMPLATE/feature_request_template.md:
```diff
@@ -4,29 +4,48 @@ about: Use this template for feature and enhancement proposals.
-### Proposed Changes
-For user-facing changes, what APIs are you proposing to add or modify? For internal changes, what code paths will need to be modified?
+Thank you for submitting a feature request. **Before proceeding, please review MLflow's
+[Issue Policy for feature requests](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md##feature-requests)
```
```suggestion
[Issue Policy for feature requests](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md#feature-requests)
```

Tiny nit, should this be a single "#"?  e.g. https://github.com/mlflow/mlflow/blob/master/CHANGELOG.rst#172-2020-03-20 works for me on chrome (jumps to 1.7.2 changelog) but https://github.com/mlflow/mlflow/blob/master/CHANGELOG.rst##172-2020-03-20 doesn't seem to work

dbczumar replied:
Yes! Thank you. Great catch!

smurching reviewed (approved):
One nit, but otherwise LGTM thanks @dbczumar!

dbczumar committed:
Update .github/ISSUE_TEMPLATE/feature_request_template.md

Co-Authored-By: Siddharth Murching <smurching@gmail.com>
